diff --git a/src/video/cocoa/SDL_cocoawindow.m b/src/video/cocoa/SDL_cocoawindow.m
index bca8eb4dd..2e444947d 100644
--- a/SDL2-src/src/video/cocoa/SDL_cocoawindow.m
+++ b/SDL2-src/src/video/cocoa/SDL_cocoawindow.m
@@ -764,6 +764,11 @@ - (void)windowDidResize:(NSNotification *)aNotification
         return;
     }

+    if (focusClickPending) {
+        focusClickPending = 0;
+        [self onMovingOrFocusClickPendingStateCleared];
+    }
+
     SDL_Window *window = _data->window;
     NSWindow *nswindow = _data->nswindow;
     int x, y, w, h;
