diff --git a/src/video/cocoa/SDL_cocoawindow.m b/src/video/cocoa/SDL_cocoawindow.m
index ad7d9e2af..f8666352a 100644
--- a/SDL2-src/src/video/cocoa/SDL_cocoawindow.m
+++ b/SDL2-src/src/video/cocoa/SDL_cocoawindow.m
@@ -783,6 +783,11 @@ - (void)windowDidResize:(NSNotification *)aNotification
         return;
     }
 
+    if (focusClickPending) {
+        focusClickPending = 0;
+        [self onMovingOrFocusClickPendingStateCleared];
+    }
+
     window = _data.window;
     nswindow = _data.nswindow;
     rect = [nswindow contentRectForFrameRect:[nswindow frame]];
